lib/pure/asynchttpserver.nim

```diff
@@ -300,9 +300,14 @@ proc processRequest(
           break
 
         # Read bytesToRead and add to body
-        # Note we add +2 because the line must be terminated by \r\n
-        let chunk = await client.recv(bytesToRead + 2)
-        request.body = request.body & chunk
+        let chunk = await client.recv(bytesToRead)
+        request.body.add(chunk)
+        # Skip \r\n (chunk terminating bytes per spec)
+        when compileOption("assertions"):
+          let separator = await client.recv(2)
+          assert separator == "\r\n"
```

Comment:
can this fail in practice? if so, it's an external condition (not internal logic), so raising exception should be used instead

silently skipping the last 2 bytes can be really bad if they're not \r\n.

I doubt checking this condition has any real performance impact compared to rest of surrounding code.

Reply:
Well, it is external input, so sure, it can trigger the assert, but it would be a malformed request according to (my understanding of) the spec.

If the bytes are never received, the `await` should never finish, so I don't think an exception would be thrown.
If the bytes are received, but are not `"\r\n"`, they simply get discarded. I also don't think this is exception worthy, especially if exceptions are explicitly disabled. 

I'm not sure (and have no strong feelings) about what the "right" thing to do if the last two bytes are not `"\r\n"`, but I would err towards saying that it is a malformed request (maybe return 400?) and is not a server exception.

Reply:
what do other languages do?

Reply:
Here's what Python does: https://github.com/python/cpython/blob/63298930fb531ba2bb4f23bc3b915dbf1e17e9e1/Lib/http/client.py#L558

```python
        if not chunk_left: # Can be 0 or None
            if chunk_left is not None:
                # We are at the end of chunk, discard chunk end
                self._safe_read(2)  # toss the CRLF at the end of the chunk
```
where `self._safe_read` is defined as
```
    def _safe_read(self, amt):
        """Read the number of bytes requested.
        This function should be used when <amt> bytes "should" be present for
        reading. If the bytes are truly not available (due to EOF), then the
        IncompleteRead exception can be used to detect the problem.
        """
        data = self.fp.read(amt)
        if len(data) < amt:
            raise IncompleteRead(data, amt-len(data))
        return data
```

Reply:
ok we can deal with it in future PR's then, keep the code as is (with `when compileOption("assertions"): ...`

Reply:
Instead of adding a TODO just do it properly: raise a ValueError.

Reply:
I agree with ValueError and that was my initial wish too in https://github.com/nim-lang/Nim/pull/16678#discussion_r555473818